File: ios/RNEmojiMainColorModule.swift
import CryptoKit
import Foundation
import React
import UIKit

@objc(RNEmojiMainColorModule)
public final class RNEmojiMainColorModule: NSObject {

  private let cacheDirectoryName = "emoji-main-color-react-native"
  private let defaultPaletteSize = 0
  private let defaultRenderSize = 64
  private let maxPaletteSize = 8
  private let minAlpha = 24
  private let minRenderSize = 16
  private let maxRenderSize = 256
  private let minWeight = 0.01
  private let mergeDistance = 42.0

  @objc
  public static func requiresMainQueueSetup() -> Bool {
    false
  }

  @objc(compute:resolver:rejecter:)
  public func compute(
    _ options: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      do {
        resolve(try self.computePayload(options: options))
      } catch {
        reject("compute_failed", "Failed to compute emoji color.", error)
      }
    }
  }

  @objc(getCachedResult:resolver:rejecter:)
  public func getCachedResult(
    _ key: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .utility).async {
      do {
        let fileURL = try self.cacheFileURL(for: key)
        guard FileManager.default.fileExists(atPath: fileURL.path) else {
          resolve(nil)
          return
        }

        let data = try Data(contentsOf: fileURL)
        let jsonObject = try JSONSerialization.jsonObject(with: data, options: [])
        resolve(jsonObject)
      } catch {
        reject("disk_cache_read_failed", "Failed to read emoji disk cache.", error)
      }
    }
  }

  @objc(setCachedResult:value:resolver:rejecter:)
  public func setCachedResult(
    _ key: String,
    value: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .utility).async {
      do {
        let fileURL = try self.cacheFileURL(for: key)
        try FileManager.default.createDirectory(
          at: fileURL.deletingLastPathComponent(),
          withIntermediateDirectories: true,
          attributes: nil
        )

        let data = try JSONSerialization.data(withJSONObject: value, options: [])
        try data.write(to: fileURL, options: .atomic)
        resolve(nil)
      } catch {
        reject("disk_cache_write_failed", "Failed to write emoji disk cache.", error)
      }
    }
  }

  private func computePayload(options: NSDictionary) throws -> NSDictionary {
    guard let rawEmoji = options["emoji"] as? String else {
      throw NSError(domain: "RNEmojiMainColorModule", code: 1)
    }

    let emoji = rawEmoji.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !emoji.isEmpty else {
      throw NSError(domain: "RNEmojiMainColorModule", code: 2)
    }

    let renderSize = clamp(
      (options["renderSize"] as? NSNumber)?.intValue ?? defaultRenderSize,
      minimum: minRenderSize,
      maximum: maxRenderSize
    )
    let paletteSize = clamp(
      (options["paletteSize"] as? NSNumber)?.intValue ?? defaultPaletteSize,
      minimum: 0,
      maximum: maxPaletteSize
    )

    let image = renderEmojiImage(emoji, renderSize: renderSize)
    let result = try analyze(image: image, paletteSize: paletteSize)
    let payload = NSMutableDictionary()
    payload["emoji"] = emoji
    payload["mainColor"] = result.mainColor
    if let palette = result.palette {
      payload["palette"] = palette
    }

    return payload
  }

  private func renderEmojiImage(_ emoji: String, renderSize: Int) -> UIImage {
    let size = CGSize(width: renderSize, height: renderSize)
    let format = UIGraphicsImageRendererFormat()
    format.opaque = false
    format.scale = 1

    let renderer = UIGraphicsImageRenderer(size: size, format: format)
    let attributedString = NSAttributedString(
      string: emoji,
      attributes: [
        .font: UIFont.systemFont(ofSize: CGFloat(renderSize) * 0.82),
      ]
    )

    return renderer.image { _ in
      UIColor.clear.setFill()
      UIRectFill(CGRect(origin: .zero, size: size))

      let bounds = attributedString.boundingRect(
        with: size,
        options: [.usesLineFragmentOrigin, .usesFontLeading],
        context: nil
      )

      let drawRect = CGRect(
        x: (size.width - bounds.width) / 2,
        y: (size.height - bounds.height) / 2 - CGFloat(renderSize) * 0.05,
        width: bounds.width,
        height: bounds.height
      ).integral

      attributedString.draw(in: drawRect)
    }
  }

  private func analyze(image: UIImage, paletteSize: Int) throws -> AnalysisResult {
    guard let cgImage = image.cgImage else {
      throw NSError(domain: "RNEmojiMainColorModule", code: 3)
    }

    let width = cgImage.width
    let height = cgImage.height
    let bytesPerRow = width * 4
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    let bitmapInfo =
      CGImageAlphaInfo.premultipliedLast.rawValue |
      CGBitmapInfo.byteOrder32Big.rawValue

    var pixels = [UInt8](repeating: 0, count: width * height * 4)
    guard let context = pixels.withUnsafeMutableBytes({ buffer in
      CGContext(
        data: buffer.baseAddress,
        width: width,
        height: height,
        bitsPerComponent: 8,
        bytesPerRow: bytesPerRow,
        space: colorSpace,
        bitmapInfo: bitmapInfo
      )
    }) else {
      throw NSError(domain: "RNEmojiMainColorModule", code: 4)
    }

    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))

    let sampleStep = max(1, width / 24)
    var buckets: [Int: Bucket] = [:]

    var index = 0
    while index < pixels.count {
      let red = Int(pixels[index])
      let green = Int(pixels[index + 1])
      let blue = Int(pixels[index + 2])
      let alpha = Int(pixels[index + 3])

      if alpha >= minAlpha {
        let weight = pixelWeight(red: red, green: green, blue: blue, alpha: alpha)
        if weight > minWeight {
          let key = quantize(red: red, green: green, blue: blue)
          let bucket = buckets[key] ?? Bucket()
          bucket.add(red: red, green: green, blue: blue, weight: weight)
          buckets[key] = bucket
        }
      }

      let pixelOffset = index / 4
      let x = pixelOffset % width
      let pixelAdvance = (x + sampleStep >= width) ? (width - x) : sampleStep
      index += pixelAdvance * 4
    }

    let weightedColors = buckets.values
      .map { $0.toWeightedColor() }
      .sorted { $0.score > $1.score }

    guard !weightedColors.isEmpty else {
      throw NSError(domain: "RNEmojiMainColorModule", code: 5)
    }

    var clusters: [Cluster] = []
    for weightedColor in weightedColors {
      if let cluster = clusters.first(where: { $0.distance(to: weightedColor) <= mergeDistance }) {
        cluster.add(weightedColor)
      } else {
        clusters.append(Cluster(weightedColor))
      }
    }

    let sortedClusters = clusters.sorted { $0.score > $1.score }
    let mainColor = sortedClusters[0].hexString()
    let palette = paletteSize > 0
      ? Array(sortedClusters.prefix(paletteSize).map { $0.hexString() })
      : nil

    return AnalysisResult(mainColor: mainColor, palette: palette)
  }

  private func pixelWeight(red: Int, green: Int, blue: Int, alpha: Int) -> Double {
    let maxChannel = Double(max(red, max(green, blue)))
    let minChannel = Double(min(red, min(green, blue)))
    let brightness = (maxChannel + minChannel) / 510.0
    let saturation = maxChannel == 0 ? 0 : (maxChannel - minChannel) / maxChannel

    var weight = (Double(alpha) / 255.0) * (0.65 + saturation * 0.7)

    if brightness > 0.95 {
      weight *= 0.25
    } else if brightness > 0.88 {
      weight *= 0.55
    }

    if brightness < 0.05 {
      weight *= 0.2
    } else if brightness < 0.12 {
      weight *= 0.45
    }

    return weight
  }

  private func quantize(red: Int, green: Int, blue: Int) -> Int {
    ((red >> 3) << 10) | ((green >> 3) << 5) | (blue >> 3)
  }

  private func clamp(_ value: Int, minimum: Int, maximum: Int) -> Int {
    min(maximum, max(minimum, value))
  }

  private func cacheFileURL(for key: String) throws -> URL {
    guard let baseURL = FileManager.default.urls(
      for: .cachesDirectory,
      in: .userDomainMask
    ).first else {
      throw NSError(domain: "RNEmojiMainColorModule", code: 6)
    }

    let digest = SHA256.hash(data: Data(key.utf8))
      .compactMap { String(format: "%02x", $0) }
      .joined()

    return baseURL
      .appendingPathComponent(cacheDirectoryName, isDirectory: true)
      .appendingPathComponent("\(digest).json", isDirectory: false)
  }
}

private final class Bucket {
  private var redTotal = 0.0
  private var greenTotal = 0.0
  private var blueTotal = 0.0
  private var weightTotal = 0.0
  private var pixelCount = 0

  func add(red: Int, green: Int, blue: Int, weight: Double) {
    redTotal += Double(red) * weight
    greenTotal += Double(green) * weight
    blueTotal += Double(blue) * weight
    weightTotal += weight
    pixelCount += 1
  }

  func toWeightedColor() -> WeightedColor {
    WeightedColor(
      red: redTotal / weightTotal,
      green: greenTotal / weightTotal,
      blue: blueTotal / weightTotal,
      weight: weightTotal,
      pixelCount: pixelCount
    )
  }
}

private struct WeightedColor {
  let red: Double
  let green: Double
  let blue: Double
  let weight: Double
  let pixelCount: Int

  var score: Double {
    weight * (1.0 + min(Double(pixelCount) / 18.0, 0.35))
  }
}

private final class Cluster {
  private var redTotal: Double
  private var greenTotal: Double
  private var blueTotal: Double
  private var weightTotal: Double
  private var pixelTotal: Int

  init(_ initialColor: WeightedColor) {
    redTotal = initialColor.red * initialColor.weight
    greenTotal = initialColor.green * initialColor.weight
    blueTotal = initialColor.blue * initialColor.weight
    weightTotal = initialColor.weight
    pixelTotal = initialColor.pixelCount
  }

  var score: Double {
    weightTotal * (1.0 + min(Double(pixelTotal) / 18.0, 0.35))
  }

  func add(_ color: WeightedColor) {
    redTotal += color.red * color.weight
    greenTotal += color.green * color.weight
    blueTotal += color.blue * color.weight
    weightTotal += color.weight
    pixelTotal += color.pixelCount
  }

  func distance(to color: WeightedColor) -> Double {
    let redDelta = averageRed - color.red
    let greenDelta = averageGreen - color.green
    let blueDelta = averageBlue - color.blue
    return sqrt((redDelta * redDelta) + (greenDelta * greenDelta) + (blueDelta * blueDelta))
  }

  func hexString() -> String {
    String(
      format: "#%02X%02X%02X",
      Int(averageRed.rounded()).clamped(to: 0...255),
      Int(averageGreen.rounded()).clamped(to: 0...255),
      Int(averageBlue.rounded()).clamped(to: 0...255)
    )
  }

  private var averageRed: Double {
    redTotal / weightTotal
  }

  private var averageGreen: Double {
    greenTotal / weightTotal
  }

  private var averageBlue: Double {
    blueTotal / weightTotal
  }
}

private struct AnalysisResult {
  let mainColor: String
  let palette: [String]?
}

private extension Int {
  func clamped(to range: ClosedRange<Int>) -> Int {
    Swift.min(range.upperBound, Swift.max(range.lowerBound, self))
  }
}

File: android/src/main/java/com/emojimaincolorreactnative/RNEmojiMainColorModule.kt
package com.emojimaincolorreactnative

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.security.MessageDigest
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

class RNEmojiMainColorModule(
  reactContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String = "RNEmojiMainColorModule"

  @ReactMethod
  fun compute(options: ReadableMap, promise: Promise) {
    try {
      val emoji = options.getString("emoji")?.trim().orEmpty()
      if (emoji.isEmpty()) {
        promise.reject("invalid_emoji", "Expected a non-empty emoji string.")
        return
      }

      val renderSize = clampInt(
        if (options.hasKey("renderSize")) options.getInt("renderSize") else DEFAULT_RENDER_SIZE,
        MIN_RENDER_SIZE,
        MAX_RENDER_SIZE,
      )
      val paletteSize = clampInt(
        if (options.hasKey("paletteSize")) options.getInt("paletteSize") else DEFAULT_PALETTE_SIZE,
        0,
        MAX_PALETTE_SIZE,
      )

      val bitmap = renderEmojiBitmap(emoji, renderSize)
      val analysis = analyzeBitmap(bitmap, paletteSize)
      promise.resolve(analysis.toWritableMap(emoji))
    } catch (error: Throwable) {
      promise.reject("compute_failed", "Failed to compute emoji color.", error)
    }
  }

  @ReactMethod
  fun getCachedResult(key: String, promise: Promise) {
    try {
      val file = cacheFile(key)
      if (!file.exists()) {
        promise.resolve(null)
        return
      }

      val payload = JSONObject(file.readText())
      promise.resolve(jsonToWritableMap(payload))
    } catch (error: Throwable) {
      promise.reject("disk_cache_read_failed", "Failed to read emoji disk cache.", error)
    }
  }

  @ReactMethod
  fun setCachedResult(key: String, value: ReadableMap, promise: Promise) {
    try {
      val payload = JSONObject()
      payload.put("emoji", value.getString("emoji"))
      payload.put("mainColor", value.getString("mainColor"))

      if (value.hasKey("palette")) {
        val palette = value.getArray("palette")
        if (palette != null) {
          payload.put("palette", readableArrayToJsonArray(palette))
        }
      }

      val file = cacheFile(key)
      file.parentFile?.mkdirs()
      file.writeText(payload.toString())
      promise.resolve(null)
    } catch (error: Throwable) {
      promise.reject("disk_cache_write_failed", "Failed to write emoji disk cache.", error)
    }
  }

  private fun renderEmojiBitmap(emoji: String, renderSize: Int): Bitmap {
    val bitmap = Bitmap.createBitmap(renderSize, renderSize, Bitmap.Config.ARGB_8888)
    val canvas = Canvas(bitmap)
    canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)

    val paint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.SUBPIXEL_TEXT_FLAG).apply {
      textAlign = Paint.Align.CENTER
      textSize = renderSize * 0.82f
    }
    val baseline = renderSize / 2f - ((paint.descent() + paint.ascent()) / 2f) - renderSize * 0.05f
    canvas.drawText(emoji, renderSize / 2f, baseline, paint)

    return bitmap
  }

  private fun analyzeBitmap(bitmap: Bitmap, paletteSize: Int): AnalysisResult {
    val width = bitmap.width
    val height = bitmap.height
    val pixels = IntArray(width * height)
    bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

    val buckets = linkedMapOf<Int, Bucket>()
    val sampleStep = max(1, width / 24)

    var index = 0
    while (index < pixels.size) {
      val pixel = pixels[index]
      val alpha = Color.alpha(pixel)
      if (alpha >= MIN_ALPHA) {
        val red = Color.red(pixel)
        val green = Color.green(pixel)
        val blue = Color.blue(pixel)
        val weight = pixelWeight(red, green, blue, alpha)

        if (weight > MIN_WEIGHT) {
          val key = quantize(red, green, blue)
          val bucket = buckets.getOrPut(key) { Bucket() }
          bucket.add(red, green, blue, weight)
        }
      }

      val x = index % width
      index += if (x + sampleStep >= width) width - x else sampleStep
    }

    if (buckets.isEmpty()) {
      throw IllegalStateException("Unable to extract a visible color from the emoji.")
    }

    val weightedColors = buckets.values
      .map { it.toWeightedColor() }
      .sortedByDescending { it.score }

    val clusters = mutableListOf<Cluster>()
    for (color in weightedColors) {
      val existingCluster = clusters.firstOrNull { it.distanceTo(color) <= MERGE_DISTANCE }
      if (existingCluster != null) {
        existingCluster.add(color)
      } else {
        clusters.add(Cluster(color))
      }
    }

    val sortedClusters = clusters.sortedByDescending { it.score() }
    val mainColor = sortedClusters.first().hex()
    val palette = if (paletteSize > 0) {
      sortedClusters.take(paletteSize).map { it.hex() }
    } else {
      null
    }

    return AnalysisResult(mainColor, palette)
  }

  private fun pixelWeight(red: Int, green: Int, blue: Int, alpha: Int): Double {
    val maxChannel = max(red, max(green, blue)).toDouble()
    val minChannel = min(red, min(green, blue)).toDouble()
    val brightness = (maxChannel + minChannel) / 510.0
    val saturation = if (maxChannel == 0.0) 0.0 else (maxChannel - minChannel) / maxChannel

    var weight = (alpha / 255.0) * (0.65 + saturation * 0.7)

    if (brightness > 0.95) {
      weight *= 0.25
    } else if (brightness > 0.88) {
      weight *= 0.55
    }

    if (brightness < 0.05) {
      weight *= 0.2
    } else if (brightness < 0.12) {
      weight *= 0.45
    }

    return weight
  }

  private fun quantize(red: Int, green: Int, blue: Int): Int {
    return ((red shr 3) shl 10) or ((green shr 3) shl 5) or (blue shr 3)
  }

  private fun jsonToWritableMap(value: JSONObject): WritableMap {
    val map = Arguments.createMap()
    map.putString("emoji", value.optString("emoji"))
    map.putString("mainColor", value.optString("mainColor"))

    if (value.has("palette")) {
      val jsonArray = value.optJSONArray("palette")
      if (jsonArray != null) {
        val palette = Arguments.createArray()
        for (index in 0 until jsonArray.length()) {
          palette.pushString(jsonArray.optString(index))
        }
        map.putArray("palette", palette)
      }
    }

    return map
  }

  private fun readableArrayToJsonArray(readableArray: ReadableArray): JSONArray {
    val jsonArray = JSONArray()

    for (index in 0 until readableArray.size()) {
      jsonArray.put(readableArray.getString(index))
    }

    return jsonArray
  }

  private fun cacheFile(key: String): File {
    val digest = MessageDigest.getInstance("SHA-256")
      .digest(key.toByteArray(Charsets.UTF_8))
      .joinToString("") { byte -> "%02x".format(byte) }

    return File(File(reactApplicationContext.cacheDir, CACHE_DIRECTORY_NAME), "$digest.json")
  }

  private fun clampInt(value: Int, minimum: Int, maximum: Int): Int {
    return min(maximum, max(minimum, value))
  }

  private data class AnalysisResult(
    val mainColor: String,
    val palette: List<String>?,
  ) {
    fun toWritableMap(emoji: String): WritableMap {
      val map = Arguments.createMap()
      map.putString("emoji", emoji)
      map.putString("mainColor", mainColor)

      if (!palette.isNullOrEmpty()) {
        val array: WritableArray = Arguments.createArray()
        palette.forEach(array::pushString)
        map.putArray("palette", array)
      }

      return map
    }
  }

  private data class WeightedColor(
    val red: Double,
    val green: Double,
    val blue: Double,
    val weight: Double,
    val pixelCount: Int,
  ) {
    val score: Double
      get() = weight * (1.0 + min(pixelCount / 18.0, 0.35))
  }

  private class Bucket {
    private var redTotal = 0.0
    private var greenTotal = 0.0
    private var blueTotal = 0.0
    private var weightTotal = 0.0
    private var pixelCount = 0

    fun add(red: Int, green: Int, blue: Int, weight: Double) {
      redTotal += red * weight
      greenTotal += green * weight
      blueTotal += blue * weight
      weightTotal += weight
      pixelCount += 1
    }

    fun toWeightedColor(): WeightedColor {
      return WeightedColor(
        red = redTotal / weightTotal,
        green = greenTotal / weightTotal,
        blue = blueTotal / weightTotal,
        weight = weightTotal,
        pixelCount = pixelCount,
      )
    }
  }

  private class Cluster(initial: WeightedColor) {
    private var redTotal = initial.red * initial.weight
    private var greenTotal = initial.green * initial.weight
    private var blueTotal = initial.blue * initial.weight
    private var weightTotal = initial.weight
    private var pixelTotal = initial.pixelCount

    fun add(color: WeightedColor) {
      redTotal += color.red * color.weight
      greenTotal += color.green * color.weight
      blueTotal += color.blue * color.weight
      weightTotal += color.weight
      pixelTotal += color.pixelCount
    }

    fun distanceTo(color: WeightedColor): Double {
      val redDelta = averageRed() - color.red
      val greenDelta = averageGreen() - color.green
      val blueDelta = averageBlue() - color.blue
      return sqrt((redDelta * redDelta) + (greenDelta * greenDelta) + (blueDelta * blueDelta))
    }

    fun score(): Double {
      return weightTotal * (1.0 + min(pixelTotal / 18.0, 0.35))
    }

    fun hex(): String {
      val red = averageRed().roundToInt().coerceIn(0, 255)
      val green = averageGreen().roundToInt().coerceIn(0, 255)
      val blue = averageBlue().roundToInt().coerceIn(0, 255)
      return String.format("#%02X%02X%02X", red, green, blue)
    }

    private fun averageRed(): Double = redTotal / weightTotal
    private fun averageGreen(): Double = greenTotal / weightTotal
    private fun averageBlue(): Double = blueTotal / weightTotal
  }

  companion object {
    private const val CACHE_DIRECTORY_NAME = "emoji-main-color-react-native"
    private const val DEFAULT_PALETTE_SIZE = 0
    private const val DEFAULT_RENDER_SIZE = 64
    private const val MAX_PALETTE_SIZE = 8
    private const val MIN_ALPHA = 24
    private const val MIN_RENDER_SIZE = 16
    private const val MAX_RENDER_SIZE = 256
    private const val MIN_WEIGHT = 0.01
    private const val MERGE_DISTANCE = 42.0
  }
}
